### ISSUE TITLE
Possible incompatibility between fixmystreet 3.0 and older versions of Debian

### ISSUE THREAD

**Issue by joesiltberg** [mysociety/fixmystreet#2972]
Posting this mostly for your information, it might not be very relevant. 

I've been trying to upgrade our beta site to fixmystreet 3.0 and ran into a problem where the pins didn't show up on the map. I've concluded that the coordinates get truncated (so a lot of the pins end up at the same point, on an integer longitude/latitude intersection, outside of the map). The coordinates were still correct in the database, but truncated in the ajax response. I spent some time debugging the perl code and it seemed to point towards the ORM layer but I didn't get very far.

In parallel, I've been working on a new installation of a newer version of Debian. We've been using Debian 8 up until now but will be switching to Debian 10. The truncation problem didn't occur in the same fixmystreet instance running on Debian 10.

Because of this I won't spend more time on it. If it's specific to Debian 8 it might not be relevant to fix since it's reaching its end of security support.

I still thought I'd let you know in case you see something similar on other systems. I guess it's not actually Debian related but perhaps related to Perl version?

**Comment by dracos** [comment]
Thanks for letting us know; strange! My guess is something locale related - truncating at integer might imply something happening with the ./, difference in the languages. But that's just a gues! It sounds like the ajax response was valid/okay, just had integers? Did it truncate new reports when you made them, or only the display of existing ones? Was the initial HTML display of a map page okay, or that affected too (I hope it'd be affected too, it should be the same code!). I can't think of anything obvious in either Debian or Perl version that could/should cause that sort of issue, though.

**Comment by joesiltberg** [comment]
Yes the ajax response seemed ok to me, it was just the coordinates that were truncated. New reports got correct coordinates in the database but didn't show up on the map (unless you zoomed out a lot and scrolled out into the Baltic sea :-) )

Not sure what you mean by "initial HTML display of a map page". Do you mean before javascript has done the ajax request? I remember seeing a pin flashing (possibly in the correct location) just after loading the map.

I guess the locale could be an explanation, but I've tried to make sure that's configured the same way on the new Debian 10 system (both in Debian, PostgreSQL and fixmystreet).

**Comment by dracos** [comment]
'Not sure what you mean by "initial HTML display of a map page"' - If you go to an `/around?lat=...&lon=...` page without `js=1` then it will show pins immediately without a JS ajax load and you can view source (or turn JS off to prevent the ajax request completely). Just to see if it was solely the ajax call I mean, which might help track it down (though probably not enough!).

**Comment by joesiltberg** [comment]
Thanks! Removing js=1 and checking viewing the source shows the correct coordinates:

```
<a title="Test från PWA" href="/report/943"><img border="0" src="/i/pin-red.png"
    class="pin js-pin" data-lat="55.672831" data-lon="13.345125"
    data-colour="red" data-id="943"
    data-title="Test från PWA" data-type=""
    data-draggable=""
    alt="Rapport" style="top:900px; left:-356px; position: absolute;"></a>
```

Then I added js=1 back and looked at the ajax response again, the element for the corresponding report under "pins" is like this:

```
[55,13,"red",943,"Test från PWA","",false]
```

The whole ajax response for completeness:

```
{"reports_list":"<ul class=\"item-list item-list--reports\" data-show-old-reports=\"1\">\n\n    \n        \n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n<li class=\"item-list__item item-list--reports__item \"\n    data-report-id=\"943\" data-lastupdate=\"2020-04-02T14:55:10\" id=\"report-943\">\n<a href=\"/report/943\">\n  \n      \n        \n        <h3 class=\"item-list__heading\">Test från PWA</h3>\n      \n  \n        <small> 14:52, torsdag  2 april 2020\n\n</small>\n\n    \n\n\n</a>\n\n</li>\n\n\n    \n        \n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n<li class=\"item-list__item item-list--reports__item \"\n    data-report-id=\"939\" data-lastupdate=\"2020-01-22T11:20:07\" id=\"report-939\">\n<a href=\"/report/939\">\n  \n      \n        \n        <h3 class=\"item-list__heading\">Test från Pixel 3</h3>\n      \n  \n        <small> 11:16, onsdag 22 januari 2020\n\n</small>\n\n    \n\n\n</a>\n\n</li>\n\n\n    \n    \n    <li class=\"item-list__item item-list__item--empty\">\n        <p>Här är några andra närliggande rapporter:</p>\n    </li>\n    \n\n</ul>\n\n\n<ul class=\"item-list item-list--reports\">\n    \n        \n        \n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n<li class=\"item-list__item item-list--reports__item \"\n    data-report-id=\"938\" data-lastupdate=\"2020-01-22T10:05:07\" id=\"report-938\">\n<a href=\"/report/938\">\n  \n        <img class=\"img\" height=\"60\" width=\"90\" src=\"/photo/938.0.fp.jpeg?797597ef\" alt=\"\">\n    \n      \n        \n        <h3 class=\"item-list__heading\">Trasig soptunna</h3>\n      \n  \n        <small> 10:02, onsdag 22 januari 2020, 9.7km\n\n</small>\n\n    \n\n\n</a>\n\n</li>\n\n\n    \n        \n        \n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n<li class=\"item-list__item item-list--reports__item \"\n    data-report-id=\"941\" data-lastupdate=\"2020-02-05T10:27:58\" id=\"report-941\">\n<a href=\"/report/941\">\n  \n      \n        \n        <h3 class=\"item-list__heading\">Testrapport från iPhone Xr</h3>\n      \n  \n        <small>\n    <span class=\"item-list__item__state\">Löst</span>\n 10:21, onsdag  5 februari 2020, 9.8km,\n      senast uppdaterad 10:27,  5 feb. 2020\n\n</small>\n\n    \n\n\n</a>\n\n</li>\n\n\n    \n</ul>\n\n","pins":[[55,13,"red",943,"Test från PWA","",false],[55,13,"red",939,"Test från Pixel 3","",false],[55,702283,13,222152,"red",938,"Trasig soptunna","",false],[55,702618,13,221077,"green",941,"Testrapport från iPhone Xr","",false]],"pagination":"\n    <p class=\"pagination\" data-page=\"1\">\n        \n\n        1 till 2 av 2\n\n        \n            <a class=\"next show_old\" href=\"https://beta.fixamingata.se/around?bbox=13.307823756104%2C55.648534017251%2C13.443436243897%2C55.736096953994&p=1&show_old_reports=1&sort=created-desc\">Visa äldre</a>\n        \n    </p>\n\n"}
```

In the new environment, the pins look fine:

```
[55.672831,13.345125,"red",943,"Test från PWA","",false]
```

**Comment by joesiltberg** [comment]
Forgot to add: I saw the pins flashing by again, and noticed that even with js=1 removed there was still an ajax call. If I disable javascript on the other hand the pins remain on the map.

**Comment by dracos** [comment]
In your full ajax reponse, 943 and 939 are truncated, but 938 and 941 are "okay" - they are using "," instead of "." which presumably means it does not work anyway, so there must be a locale issue in there somewhere. But I don't understand why some entries would truncate and some would output (but in Swedish locale), sorry.

**Comment by joesiltberg** [comment]
Ahh, well spotted!

I'll try to spend some more time with the perl code. Maybe there's something with the JSON serialization...I'm thinking that if there's a locale issue _before_ the pin array reaches the JSON serialization, the coordinate would be a string by then and serialized as "1,23" rather than 1,23.

**Comment by joesiltberg** [comment]
Here's some more information:

The problems on the map get truncated already when getting them from map_features() in FixMyStreet::Map. The others (nearby) are correct at first, but then incorrectly serialized in encode_json.

The difference between the on_map problems and the nearby problems seem to be that we're wrapping the db query with mySociety::Locale::in_gb_locale in different ways. For nearby the call to ResultSet::all (which I guess executes the query) is included in the in_gb_locale-block, whereas for the on_map problems all is called outside of the in_gb_locale block (see around_map in FixMyStreet/DB/ResultSet/Problem.pm and map_features in FixMyStreet/Map.pm).

If I add the in_gb_locale-thing around the call to all in map_features in FixMyStreet/Map.pm, and also around encode_json in ajax() in FixMyStreet/App/Controller/Reports.pm, the pins show up correctly again.

I still have no idea why this wasn't needed before the upgrade to 3.0, or why it's not needed on my Debian 10 server.
